packages/flutter/lib/src/material/checkbox.dart

```diff
@@ -470,6 +469,17 @@ class _CheckboxState extends State<Checkbox> with TickerProviderStateMixin, Togg
 
   @override
   Widget build(BuildContext context) {
+    final CheckboxThemeData checkboxTheme = CheckboxTheme.of(context);
+
+    // Colors need to be resolved in selected and non selected states separately
+    final Set<MaterialState> activeStates = states..add(MaterialState.selected);
+    final Set<MaterialState> inactiveStates = states..remove(MaterialState.selected);
+    if (widget.isError) {
+      activeStates.add(MaterialState.error);
+      inactiveStates.add(MaterialState.error);
+    }
+    final Set<MaterialState> checkStates = widget.isError ? (states..add(MaterialState.error)) : states;
```

Comment:
Does this really work? Now that `CupertinoCheckbox` is returned, I would expect that `Checkbox.states` is not updated at all.

Reply:
I'm not sure I understand the question. I think this line is just adding `WidgetState.error` to a copy of `states` which is then used to resolve the check color in different `WidgetState`s.

Reply:
The `states` here belong to `Checkbox`, not `CupertinoCheckbox`, and will not be updated by many internal logics. States such as "selected" or "disabled" might work, but "hover" won't work.

I've built this reproduction code:

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      // theme: Theme.of(context).copyWith(checkboxTheme: const CheckboxThemeData(mouseCursor: MyMouseCursor())),
      home: CupertinoPageScaffold(
        navigationBar: CupertinoNavigationBar(
          middle: Text('CupertinoCheckbox Example'),
        ),
        child: SafeArea(
          child: CupertinoCheckboxExample(),
        ),
      ),
    );
  }
}

class CupertinoCheckboxExample extends StatefulWidget {
  const CupertinoCheckboxExample({super.key});

  @override
  State<CupertinoCheckboxExample> createState() => _CupertinoCheckboxExampleState();
}

class _CupertinoCheckboxExampleState extends State<CupertinoCheckboxExample> {
  bool? isChecked = true;

  @override
  Widget build(BuildContext context) {
    return Checkbox.adaptive(
    // return CupertinoCheckbox(
      mouseCursor: const MyMouseCursor(),
      value: isChecked,
      onChanged: (bool? value) {
        setState(() {
          isChecked = value;
        });
      },
    );
  }
}

class MyMouseCursor extends WidgetStateMouseCursor {
  const MyMouseCursor();

  @override
  MouseCursor resolve(Set<WidgetState> states) {
    if (states.contains(WidgetState.hovered)) {
      return SystemMouseCursors.forbidden;
    }
    return SystemMouseCursors.basic;
  }

  @override
  String get debugDescription => 'My';
}
```

In this code, `MyMouseCursor` shows "forbidden" cursor if when the widget is hovered (which means it should always show "forbidden", since mouse cursors only make sense when hovered).

If you pass in `MyMouseCursor()` using `CheckBox.adaptive`'s constructor directly, then the mouse cursors is correctly "forbidden", because your current code pass in the unresolved WidgetStateCursor.

However, if you pass in `MyMouseCursor()` using the theme data, you'll find that the cursor is not correctly set. This is because the `state` you used to resolve the theme cursor with is not updated by the actual gesture handlers within `CupertinoCheckBox`. You can log the `state` here and within `_CupertinoCheckBoxState` and notice that the former is never updated when hovered.

In short, anything passed into `CupertinoCheckbox` here should not be resolved, because this `state` is not reliable.

Reply:
I see. Thanks for the sample code, it was really easy to reproduce.

I think I missed this issue because in my tests I was only checking that `CupertinoCheckbox` could get values from `CheckboxThemeData`, I should have also checked that `states` was also being resolved properly. 

As you said, `side: widget.side ?? checkboxTheme.side` would work but `mouseCursor: widget.mouseCursor ?? checkboxTheme.mouseCursor?.resolve(states)` would not.

Would passing the theme property into `CupertinoCheckbox` solve this problem? So instead of 

```dart
mouseCursor: widget.mouseCursor ?? checkboxTheme.mouseCursor?.resolve(states)
``` 

do 

```dart
mouseCursor: widget.mouseCursor ?? checkboxTheme.mouseCursor
```

and then resolve in `CupertinoCheckbox`'s own states? But then that would require changing all `CupertinoCheckbox` properties to be of exactly the same type as their corresponding `CheckboxThemeData` properties? I think @justinmc raised an alarm about this earlier when he was talking about the type mismatch of `checkColor`. 

Alternatively is there some other solution?

This is probably a useful discussion even outside the scope of this PR and in the overall adaptive effort.